extensions/amp-sidebar/1.0/sidebar.js

```diff
@@ -0,0 +1,241 @@
+import {ContainWrapper} from '../../../src/preact/component';
+import {assertDoesNotContainDisplay, setStyles} from '../../../src/style';
+import {forwardRef} from '../../../src/preact/compat';
+import {
+  useCallback,
+  useImperativeHandle,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './sidebar.jss';
+
+const ANIMATION_DURATION = 350;
+const ANIMATION_EASE_IN = 'cubic-bezier(0,0,.21,1)';
+
+const ANIMATION_KEYFRAMES_FADE_IN = [{'opacity': '0'}, {'opacity': '1'}];
+const ANIMATION_KEYFRAMES_SLIDE_IN_LEFT = [
+  {'transform': 'translateX(-100%)'},
+  {'transform': 'translateX(0)'},
+];
+const ANIMATION_KEYFRAMES_SLIDE_IN_RIGHT = [
+  {'transform': 'translateX(100%)'},
+  {'transform': 'translateX(0)'},
+];
+
+const ANIMATION_STYLES_SIDEBAR_LEFT_INIT = {'transform': 'translateX(-100%)'};
+const ANIMATION_STYLES_SIDEBAR_RIGHT_INIT = {'transform': 'translateX(100%)'};
+const ANIMATION_STYLES_MASK_INIT = {'opacity': '0'};
+const ANIMATION_STYLES_SIDEBAR_FINAL = {'transform': ''};
+const ANIMATION_STYLES_MASK_FINAL = {'opacity': ''};
+
+/**
+ * @param {T} current
+ * @return {{current: T}}
+ * @template T
+ */
+function useValueRef(current) {
+  const valueRef = useRef(null);
+  valueRef.current = current;
+  return valueRef;
+}
+/**
+ * @param {!Element} element
+ * @param {!Object<string, *>} styles
+ */
+function safelySetStyles(element, styles) {
+  setStyles(element, assertDoesNotContainDisplay(styles));
+}
+
+/**
+ * @param {!SidebarDef.Props} props
+ * @param {{current: (!SidebarDef.SidebarApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function SidebarWithRef(
+  {
+    as: Comp = 'div',
+    side = 'left',
+    onBeforeOpen,
+    onAfterClose,
+    children,
+    ...rest
+  },
+  ref
+) {
+  // There are two phases to open and close.
+  // To open, we mount and render the contents (invisible), then animate the display (visible).
+  // To close, it's the reverse.
+  // `mounted` mounts the component. `opened` plays the animation.
+  const initialOpen = false;
+  const [mounted, setMounted] = useState(initialOpen);
+  const [opened, setOpened] = useState(initialOpen);
+  const classes = useStyles();
+  const sidebarRef = useRef();
+  const maskRef = useRef();
+
+  // We are using refs here to refer to common strings, objects, and functions used.
+  // This is because they are needed within `useEffect` calls below (but are not depended for triggering)
+  // We use `useValueRef` for props that might change (user-controlled)
+  const onBeforeOpenRef = useValueRef(onBeforeOpen);
+  const onAfterCloseRef = useValueRef(onAfterClose);
+
+  const open = useCallback(() => {
+    if (onBeforeOpenRef.current) {
+      onBeforeOpenRef.current();
+    }
+    setMounted(true);
+    setOpened(true);
+  }, [onBeforeOpenRef]);
+  const close = useCallback(() => setOpened(false), []);
+  const toggle = useCallback(() => (opened ? close() : open()), [
+    opened,
```

Comment:
Would it be helpful to have an `openedRef` so this callback value does not have to be redefined every time the state changes?

Reply:
Thanks, will address this as a follow up once we get further into the sidebar implementation.